specs/emip-005.md

```diff
@@ -0,0 +1,472 @@
+
+### Structure
+
+The COSE specification is made to be composable -- that is you can have a plaintext that you wrap with a signature, then wrap with an encryption, then wrap with a signature again (and so on).
+
+That means that for encryption in particular, it can either
+
+1) Be used to encrypt plaintext directly
+2) Be used to encrypt another COSE message
+
+In this spec, we care about the case where you encrypt a `signed_message`
+
+Here is the overall CBOR structure
+
+```
+# holds encrypted content itself
+COSE_Encrypt = [
+  Headers,
+  ciphertext : bstr / nil, # contains encrypted signed_message
+  recipients : [+COSE_recipient]
+]
+
+# holds encrypted keys the receiver can use to decrypt the content
+COSE_recipient = [
+  Headers,
+  ciphertext : bstr / nil, # contains encrypted key to decrypt the COSE_Encrypt ciphertext
+  ? recipients : [+COSE_recipient] # in case you need multiple rounds to get decryption key
+]
+```
+
+To encrypt, the structure as a whole, we call our encryption method once for each level (root, recipient, etc.) recursively. For example, we encrypt the `signed_message` and put it in the `COSE_Encrypt` ciphertext, then we encrypt the decryption key and put it in the `COSE_recipient` ciphertext.
+
+For the `Headers`,
+
+- The `protected` fields MUST be empty. These are meant to be used with AEAD which we  don't need in this  specification (you can read more about it at [RFC 8152 section 5.3](https://tools.ietf.org/html/rfc8152#section-5.3)).
+- The `unprotected` field MUST contain a `enc_type` label with value `EncryptionType` so that wallets can easily know what UI to display to users.
+
+We define the encrypted payload as the union of two possibilities (defined below)
+
+```
+encrypted_message = PasswordEncryption / PubKeyEncryption
+```
+
+#### Password-based encryption
+
+For password-based encryption we don't need a receiver field (anybody who knows the password can decrypt) so we instead use the following (simplified) structure
+
+```
+COSE_Encrypt0 = [
+    Headers,
+    ciphertext : bstr / nil,
+]
+
+PasswordEncryption = 16 (COSE_Encrypt0)
+```
+
+The `Headers` MUST have `enc_type: 1` in the `unprotected` field.
+
+The `ciphertext` is simply the result of [the Yoroi encryption spec](https://github.com/Emurgo/yoroi-frontend/blob/737595fec5a89409aacef827d356c9a1605515c0/docs/specs/code/ENCRYPT.md) on the `signed_message`.
+
+#### Public key based encryption
+
+We only allow encrypting based on `ED25519 extended` public keys (the ones used for Cardano). To encrypt based on these public keys, you must
+
+1) Compute a password consisting of 22 case-sensitive alphanumeric (a–z, A–Z, 0–9) characters (this gives you ~128 bits of entropy)
+
+Now, for each receiver, you must
+
+2) Compute an ephemeral key pair using `ED255519 extended`
+1) Compute the shared DH secret between the private key from step (1) and the public key received (using the `exchange` functionality)
```

Comment:
Is there a case where picking a specific ephemeral key pair makes the shared secret reveal information about a private key? If so, you could for example use this to extract private keys out of your hardware wallet

Reply:
I think it's not an issue, since next step of the protocol is a Blake2b512 hashing, which will safely "gather all the entropy" of obtained shared secret and spread it around as needed

Reply:
> What is very important - an adversary should not be able to alter a public key which is received

could you clarify this point? Right now `eph_pub` is unprotected so a man-in-the-middle could modify it before it arrives. If you use the  wrong `eph_pub` does the message not  just fail to decrypt? (Encryption spec can detect bad passwords). Making anything after the encryption step "protected" would be awkward because then for a given message we would have to

1) Sign the message M which results in a signed message S
2) Encrypt signed message S which results in an encrypted message M
3) Sign the `eph_pub` in M? Which key would we sign with? (don't want to reveal the public key used in step (1) since that would partially reveal information).

Requiring a step (3) would also be poor UX because for hardware wallet users you would have to
1) Sign the message with the hardware wallet (requires user accepting a prompt on the hardware wallet itself)
2) Send S back to the desktop and then the computer generates M
3) Send M to the hardware wallet so it can sign again (so the user now needs to  be prompted again)

Reply:
I mean that public key of memo recipient have to be protected against forgery (not `eph_pub` - decryption will really be failed). 

If I understand correctly, when A wants to send encrypted memo to B, A and B should both have memo key pair chain (as defined in EmiP4). Most likely, A's and B's memo chains will have different hierarchies. For new memo encrypting, A will derive new key pair (based on memo chain) specific to the `header index` of the memo. Then A takes account (master) public key of B and performs memo encryption. It's critical for A to take valid public key that really belongs to B.